File: timers.ts
// Copyright 2018 Ryan Dahl <ry@tinyclouds.org>
// All rights reserved. MIT License.
import { main as pb } from "./msg.pb";
import * as dispatch from "./dispatch";
import { assert } from "./util";

let nextTimerId = 1;

// tslint:disable-next-line:no-any
export type TimerCallback = (...args: any[]) => void;

interface Timer {
  id: number;
  cb: TimerCallback;
  interval: boolean;
  // tslint:disable-next-line:no-any
  args: any[];
  delay: number; // milliseconds
}

const timers = new Map<number, Timer>();

export function initTimers() {
  dispatch.sub("timers", onMessage);
}

function onMessage(payload: Uint8Array) {
  const msg = pb.Msg.decode(payload);
  assert(msg.command === pb.Msg.Command.TIMER_READY);
  const id = msg.timerReadyId;
  const done = msg.timerReadyDone;
  const timer = timers.get(id);
  if (!timer) {
    return;
  }
  timer.cb(...timer.args);
  if (done) {
    timers.delete(id);
  }
}

function setTimer(
  cb: TimerCallback,
  delay: number,
  interval: boolean,
  // tslint:disable-next-line:no-any
  args: any[]
): number {
  const timer = {
    id: nextTimerId++,
    interval,
    delay,
    args,
    cb
  };
  timers.set(timer.id, timer);
  dispatch.sendMsg("timers", {
    command: pb.Msg.Command.TIMER_START,
    timerStartId: timer.id,
    timerStartInterval: timer.interval,
    timerStartDelay: timer.delay
  });
  return timer.id;
}

export function setTimeout(
  cb: TimerCallback,
  delay: number,
  // tslint:disable-next-line:no-any
  ...args: any[]
): number {
  return setTimer(cb, delay, false, args);
}

export function setInterval(
  cb: TimerCallback,
  delay: number,
  // tslint:disable-next-line:no-any
  ...args: any[]
): number {
  return setTimer(cb, delay, true, args);
}

export function clearTimer(id: number) {
  dispatch.sendMsg("timers", {
    command: pb.Msg.Command.TIMER_CLEAR,
    timerClearId: id
  });
}

File: timers.go
// Copyright 2018 Ryan Dahl <ry@tinyclouds.org>
// All rights reserved. MIT License.
package deno

import (
	"github.com/golang/protobuf/proto"
	"time"
)

type Timer struct {
	Id       int32
	Done     bool
	Cleared  bool
	Interval bool
	Duration int32 // In milliseconds
}

var timers = make(map[int32]*Timer)

func InitTimers() {
	Sub("timers", func(buf []byte) []byte {
		msg := &Msg{}
		check(proto.Unmarshal(buf, msg))
		switch msg.Command {
		case Msg_TIMER_START:
			id := msg.TimerStartId
			t := &Timer{
				Id:       id,
				Done:     false,
				Interval: msg.TimerStartInterval,
				Duration: msg.TimerStartDuration,
				Cleared:  false,
			}
			// If this parameter is less than 10, a value of 10 is used
			if t.Duration < 10 {
				t.Duration = 10
			}
			t.StartTimer()
			timers[id] = t
			return nil
		case Msg_TIMER_CLEAR:
			// TODO maybe need mutex here.
			timer := timers[msg.TimerClearId]
			timer.Clear()
		default:
			panic("[timers] Unexpected message " + string(buf))
		}
		return nil
	})
}

func (t *Timer) Clear() {
	if !t.Cleared {
		wg.Done()
		t.Cleared = true
		delete(timers, t.Id)
	}
	t.Done = true
}

func (t *Timer) StartTimer() {
	wg.Add(1)
	go func() {
		defer t.Clear()
		for {
			time.Sleep(time.Duration(t.Duration) * time.Millisecond)
			if !t.Interval {
				t.Done = true
			}
			PubMsg("timers", &Msg{
				Command:        Msg_TIMER_READY,
				TimerReadyId:   t.Id,
				TimerReadyDone: t.Done,
			})
			if t.Done {
				return
			}
		}
	}()
}
